crates_io_cdn_logs/src/lib.rs

```diff
@@ -0,0 +1,202 @@
+pub mod cloudfront;
+mod compression;
+mod download_map;
+pub mod fastly;
+mod paths;
+#[cfg(test)]
+mod test_utils;
+
+pub use crate::compression::Decompressor;
+pub use crate::download_map::DownloadsMap;
+use std::io::Cursor;
+use tokio::io::{AsyncBufRead, AsyncReadExt};
+
+pub async fn count_downloads<R>(mut reader: R) -> anyhow::Result<DownloadsMap>
+where
+    R: AsyncBufRead + Unpin,
+{
+    // Read the first byte to determine the file format.
+    match reader.read_u8().await? {
```

Comment:
I'm handling my notification backlog in chronological order, so there's probably context in a later PR or issue that explains this, but is sniffing something we actually need here? Is there a scenario where we won't know the source of a log file?

(Personal bias: I'm automatically mistrustful of anything that looks like data sniffing.)

Reply:
the alternative is looking at the folder structure on the S3 bucket and choosing the file type based on that. or the other alternative: fastly logs are using zstd and cloudfront is currently using gzip.

both of those approaches feel a bit brittle though, which is why I implemented the auto-detection.